src/SprykerEco/Yves/Heidelpay/Processor/HeidelpayDirectDebitRegistrationProcessor.php

```diff
@@ -0,0 +1,244 @@
+                static::ERROR_CODE_REGISTRATION_NOT_FOUND,
+                $directDebitRegistrationTransfer
+            );
+        }
+
+        $quoteTransfer = $this->addDirectDebitRegistrationToQuote($directDebitRegistrationTransfer, $quoteTransfer);
+        $this->quoteClient->setQuote($quoteTransfer);
+
+        return $directDebitRegistrationTransfer;
+    }
+
+    /**
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     *
+     * @return \Generated\Shared\Transfer\HeidelpayDirectDebitRegistrationTransfer
+     */
+    protected function createNewDirectDebitRegistrationTransferFromRequest(Request $request): HeidelpayDirectDebitRegistrationTransfer
+    {
+        $apiResponseAsArray = $this->getUrldecodedRequestBody($request);
+        $apiResponseAsArray = $this->filterResponseParameters($apiResponseAsArray);
+
+        return $this->heidelpayClient->parseDirectDebitRegistrationResponse($apiResponseAsArray);
+    }
+
+    /**
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     * @param \Generated\Shared\Transfer\QuoteTransfer $quoteTransfer
+     *
+     * @return \Generated\Shared\Transfer\HeidelpayDirectDebitRegistrationTransfer
+     */
+    protected function retrieveExistingDirectDebitRegistrationByRequestAndQuote(
+        Request $request,
+        QuoteTransfer $quoteTransfer
+    ): HeidelpayDirectDebitRegistrationTransfer {
+        $directDebitRegistrationTransfer = (new HeidelpayDirectDebitRegistrationTransfer())
+            ->setRegistrationUniqueId($this->getRegistrationId($request))
+            ->setIdCustomerAddress($quoteTransfer->getShippingAddress()->getIdCustomerAddress())
+            ->setTransactionId($this->generateTransactionId($quoteTransfer));
+
+        return $directDebitRegistrationTransfer;
+    }
+
+    /**
+     * @param string $code
+     * @param \Generated\Shared\Transfer\HeidelpayDirectDebitRegistrationTransfer $directDebitRegistrationTransfer
+     *
+     * @return \Generated\Shared\Transfer\HeidelpayDirectDebitRegistrationTransfer
+     */
+    protected function addError(string $code, HeidelpayDirectDebitRegistrationTransfer $directDebitRegistrationTransfer): HeidelpayDirectDebitRegistrationTransfer
+    {
+        $directDebitRegistrationTransfer
+            ->setIsError(true)
+            ->setError(
+                (new HeidelpayResponseErrorTransfer())
+                    ->setCode($code)
+            );
+
+        return $directDebitRegistrationTransfer;
+    }
+
+    /**
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     *
+     * @return int
+     */
+    protected function getRegistrationId(Request $request): int
+    {
+        return (int)$request->get(static::REQUEST_PARAM_REGISTRATION_ID);
+    }
+
+    /**
+     * @param \Generated\Shared\Transfer\QuoteTransfer $quoteTransfer
+     *
+     * @return bool
+     */
+    protected function isQuoteExpired(QuoteTransfer $quoteTransfer): bool
+    {
+        try {
```

Comment:
I would prefer not to use "Exception driven development". You can have a condition here